django_migration_linter/sql_analyser/postgresql.py

```diff
@@ -5,6 +5,36 @@
 from .base import BaseAnalyser, Check, CheckMode, CheckType
 
 
+def has_create_index_in_transaction(sql_statements: list[str], **kwargs) -> bool:
+    """Return if a migration opens a transaction, acquires EXCLUSIVE lock, then indexes.
+
+    Any locks that are obtained after a transaction is opened will not be released
+    until that transaction either commits or rolls back.
+
+    Accordingly, it's extremely important that a migration avoid any long-running
+    queries while exclusive locks are held.
+
+    Building an index is generally a long-running operation. A standard index
+    creation (that is, avoiding use of `CONCURRENTLY`) can be safe to run on
+    tables which are small in size and/or read-heavy. (`CREATE INDEX` only
+    prevents writes to a table; reads are allowed during index creation).
+
+    This check is a stricter version of `CREATE_INDEX` -- if a team wishes
+    to build indices nonconcurrently, it's imperative to be mindful of locks.
+    """
+    if not (sql_statements and sql_statements[0].startswith("BEGIN")):
+        return False
+
+    for i, sql in enumerate(sql_statements):
+        # If any statements acquire an exclusive lock, complain about index creation later.
+        # Nearly every single `ALTER TABLE` command requires an exclusive lock:
+        #     https://www.postgresql.org/docs/current/sql-altertable.html
+        # (Most common example is `ALTER TABLE... ADD COLUMN`, then later `CREATE INDEX`)
+        if sql.startswith("ALTER TABLE"):
```

Comment:
This check may give false positives for operations, that do not take long time, like dropping columns, indexes or anything else. I suggest to explicitly specify the list of potentially long operations.
```suggestion
        long_operations = ("ADD COLUMN", "CREATE INDEX")
        column_or_index_creation = any(op in sql for op in long_operations)
        if sql.startswith("ALTER TABLE") and column_or_index_creation:
```

Reply:
As an example, consider the following SQL:
```sql
BEGIN;
ALTER TABLE "users" DROP COLUMN "last_name";  -- fast, EXCLUSIVE
CREATE INDEX "user_email" ON "users" ("email");  -- slow, SHARED
COMMIT;
```

Even though the `DROP COLUMN` should be very fast, the exclusive lock it requires is held until the indexing operation releases.

The problem goes away if you either:
- stop using transactions
- remove the `ALTER TABLE` that comes before index creation

Reply:
Ok, David, thank you for explanation. This is one of the things I did not know about Postgesql. I am ok with this check. What however I find missing, is more informative error message, and the most important - ways to resolve it. Please, improve the error message, so that it not only explains the problem, but also how to fix it. Something like: you need to either disable transaction by adding xxx, or split this migration in 2 parts to separate expensive operations.

Reply:
That's an excellent suggestion, and I'll make efforts to improve the clarity here. Thank you for your feedback!